src/board/v2/nap/track_channel.c

```diff
@@ -62,34 +70,38 @@ void nap_track_init_pack(u8 pack[], u8 prn, s32 carrier_phase, u16 code_phase)
  * \param carrier_phase Initial code phase.
  * \param code_phase    Initial carrier phase.
  */
-void nap_track_init_wr_blocking(u8 channel, u8 prn, s32 carrier_phase,
-                                u16 code_phase)
+void nap_track_init(u8 channel, gnss_signal_t sid, u32 ref_timing_count,
+                    float carrier_freq, float code_phase)
 {
-  u8 temp[6] = { 0, 0, 0, 0, 0, 0 };
-
-  nap_track_init_pack(temp, prn, carrier_phase, code_phase);
-  nap_xfer_blocking(NAP_REG_TRACK_BASE + channel * NAP_TRACK_N_REGS
-                     + NAP_REG_TRACK_INIT_OFFSET, 6, 0, temp);
+  struct nap_ch_state *s = &nap_ch_state[channel];
+  memset(s, 0, sizeof(*s));
+
+  u32 track_count = nap_timing_count() + 20000;
+  float cp = propagate_code_phase(code_phase, carrier_freq,
+                                  track_count - ref_timing_count);
+
+  /* Contrive for the timing strobe to occur at or close to a
+   * PRN edge (code phase = 0) */
+  track_count += (SAMPLE_FREQ/GPS_CA_CHIPPING_RATE) * (1023.0-cp) *
+                 (1.0 + carrier_freq / GPS_L1_HZ);
+
+  nap_track_code_wr_blocking(channel, sid);
+  nap_track_init_wr_blocking(channel, 0, 0, 0);
+
+  double cp_rate = (1 + carrier_freq/GPS_L1_HZ) * GPS_CA_CHIPPING_RATE;
+  nap_track_update(channel, carrier_freq, cp_rate, 0, 0);
+
+  /* Schedule the timing strobe for start_sample_count. */
+  track_count -= SAMPLE_FREQ / (2*GPS_CA_CHIPPING_RATE);
+  nap_timing_strobe(track_count);
+  nap_timing_strobe_wait(100);
+  
+  s->count_snapshot = track_count;
```

Comment:
Need to ensure that these are stored before before the first interrupt occurs and `nap_track_update()` is called. Probably want
1. Writes to `s`
2. Compiler barrier
3. `nap_timing_strobe()`